libs/counterfactuals/src/lib/CounterfactualChartWithLegend.tsx

```diff
@@ -0,0 +1,189 @@
+          <CounterfactualChartLegend
+            {...this.props}
+            customPointIsActive={this.state.customPointIsActive}
+            customPoints={this.state.customPoints}
+            selectedPointsIndexes={this.props.selectedPointsIndexes}
+            removeCustomPoint={this.removeCustomPoint}
+            setTemporaryPointToCopyOfDatasetPoint={
+              this.props.setTemporaryPointToCopyOfDatasetPoint
+            }
+            toggleCustomActivation={this.toggleCustomActivation}
+            togglePanel={this.togglePanel}
+            toggleSelectionOfPoint={this.toggleSelectionOfPoint}
+          />
+        </Stack>
+      </Stack.Item>
+    );
+  }
+
+  private toggleSelectionOfPoint = (index?: number): void => {
+    if (index === undefined) {
+      return;
+    }
+    const indexOf = this.props.selectedPointsIndexes.indexOf(index);
+    let newSelections = [...this.props.selectedPointsIndexes];
+    let pointIsActive = [...this.state.pointIsActive];
+    let originalData;
+    if (indexOf === -1) {
+      newSelections = [index];
+      pointIsActive = [true];
+      originalData = getOriginalData(
+        index,
+        this.context.jointDataset,
+        this.context.dataset
+      );
+    } else {
+      newSelections.splice(indexOf, 1);
+      pointIsActive.splice(indexOf, 1);
+    }
+    this.setState({
+      originalData,
+      pointIsActive
+    });
+    this.props.onSelectedPointsIndexesUpdated(newSelections);
+  };
+
+  private toggleCustomActivation = (index: number): void => {
+    const customPointIsActive = [...this.state.customPointIsActive];
+    customPointIsActive[index] = !customPointIsActive[index];
+    this.setState({ customPointIsActive });
+  };
+
+  private removeCustomPoint = (index: number): void => {
+    this.setState((prevState) => {
+      const customPoints = [...prevState.customPoints];
+      customPoints.splice(index, 1);
+      const customPointIsActive = [...prevState.customPointIsActive];
+      customPointIsActive.splice(index, 1);
+      this.props.onCustomPointLengthUpdated(customPoints.length);
+      return { customPointIsActive, customPoints };
+    });
+  };
+
+  private saveAsPoint = (): void => {
+    const customPoints = [...this.state.customPoints];
+    const customPointIsActive = [...this.state.customPointIsActive];
+    if (this.props.temporaryPoint) {
+      customPoints.push(this.props.temporaryPoint);
+    }
+    customPointIsActive.push(true);
+    this.setState({
+      customPointIsActive,
+      customPoints
+    });
+    this.props.onCustomPointLengthUpdated(customPoints.length);
+  };
+
+  private togglePanel = (): void => {
+    if (!this.state.isPanelOpen) {
+      this.logTelemetryEvent(
+        TelemetryEventName.CounterfactualCreateWhatIfCounterfactualClick
+      );
+    }
+    this.setState((preState) => {
+      return { isPanelOpen: !preState.isPanelOpen };
+    });
+  };
+
+  private logTelemetryEvent = (eventName: TelemetryEventName) => {
```

Comment:
I see this method copied several times, should it live in some global shared utils?

Reply:
Great comment! Yeah there is already a work item to track it: [Task 1893214](https://msdata.visualstudio.com/Vienna/_workitems/edit/1893214): Create a context for telemetryHook